File: WhisperApp/WhisperApp/Services/LLMService.swift
// Copyright (c) 2026 Roman Barinov. MIT License.

import Foundation

enum LLMError: LocalizedError {
    case noAPIKey
    case invalidURL
    case networkError(Error)
    case apiError(String)
    case decodingError

    var errorDescription: String? {
        switch self {
        case .noAPIKey: return "API key is not configured. Open Settings to set it."
        case .invalidURL: return "Invalid API base URL. Check Settings."
        case .networkError(let error): return "Network error: \(error.localizedDescription)"
        case .apiError(let message): return "API error: \(message)"
        case .decodingError: return "Failed to parse LLM response."
        }
    }
}

class LLMService {
    static let shared = LLMService()

    private struct ChatCompletionResponse: Decodable {
        struct Choice: Decodable {
            struct Message: Decodable {
                let content: String
            }
            let message: Message
        }
        let choices: [Choice]
    }

    private let maxRetries = 3
    private let retryDelays: [UInt64] = [500_000_000, 1_500_000_000, 3_000_000_000] // 0.5s, 1.5s, 3s

    func process(text: String) async throws -> String {
        let settings = AppSettings.shared

        guard !settings.apiKey.isEmpty else { throw LLMError.noAPIKey }

        let baseURL = settings.apiBaseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard let url = URL(string: "\(baseURL)/v1/chat/completions") else {
            throw LLMError.invalidURL
        }

        // Build messages array
        var messages: [[String: String]] = []
        let systemPrompt = settings.llmSystemPrompt.trimmingCharacters(in: .whitespacesAndNewlines)
        if !systemPrompt.isEmpty {
            messages.append(["role": "system", "content": systemPrompt])
        }
        // Wrap transcription in tags to clearly separate data from instructions
        messages.append(["role": "user", "content": "<transcription>\(text)</transcription>"])

        let requestBody: [String: Any] = [
            "model": settings.llmModelName,
            "messages": messages,
            "reasoning_effort": "low"
        ]

        guard let bodyData = try? JSONSerialization.data(withJSONObject: requestBody) else {
            throw LLMError.decodingError
        }

        var lastError: Error = LLMError.networkError(
            NSError(domain: "WhisperApp", code: -1, userInfo: [NSLocalizedDescriptionKey: "Unknown error"])
        )

        for attempt in 0..<maxRetries {
            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.setValue("Bearer \(settings.apiKey)", forHTTPHeaderField: "Authorization")
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.timeoutInterval = 30
            request.httpBody = bodyData

            do {
                let (data, response) = try await URLSession.shared.data(for: request)

                if let httpResponse = response as? HTTPURLResponse {
                    if httpResponse.statusCode == 200 {
                        guard let result = try? JSONDecoder().decode(ChatCompletionResponse.self, from: data),
                              let content = result.choices.first?.message.content else {
                            throw LLMError.decodingError
                        }
                        return content.trimmingCharacters(in: .whitespacesAndNewlines)
                    }

                    let errorBody = String(data: data, encoding: .utf8) ?? "Unknown error"
                    let apiErr = LLMError.apiError("HTTP \(httpResponse.statusCode): \(errorBody)")

                    // Don't retry client errors (4xx) except 408/429
                    if httpResponse.statusCode >= 400 && httpResponse.statusCode < 500
                        && httpResponse.statusCode != 408 && httpResponse.statusCode != 429 {
                        throw apiErr
                    }
                    lastError = apiErr
                }
            } catch let error as LLMError {
                // Non-retryable errors
                switch error {
                case .noAPIKey, .invalidURL, .decodingError:
                    throw error
                case .apiError:
                    // 4xx already thrown above; if we get here it's a retryable apiError
                    lastError = error
                case .networkError:
                    lastError = error
                }
            } catch {
                lastError = LLMError.networkError(error)
            }

            // Wait before retrying (unless last attempt)
            if attempt < maxRetries - 1 {
                try? await Task.sleep(nanoseconds: retryDelays[attempt])
            }
        }

        // All retries exhausted
        throw lastError
    }
}

File: electron-app/src/main/services/llm-service.ts
// Copyright (c) 2026 Roman Barinov. MIT License.

import * as http from 'http';
import * as https from 'https';
import { LLM_TIMEOUT_MS } from '../../shared/constants';
import { retryWithBackoff } from './retry';

export class LLMError extends Error {
  constructor(
    message: string,
    public readonly code:
      | 'NO_API_KEY'
      | 'INVALID_URL'
      | 'NETWORK_ERROR'
      | 'API_ERROR'
      | 'DECODING_ERROR',
    public readonly statusCode?: number
  ) {
    super(message);
    this.name = 'LLMError';
  }
}

export interface LLMSettings {
  apiKey: string;
  apiBaseURL: string;
  llmModelName: string;
  llmSystemPrompt: string;
}

export interface RequestOptions {
  signal?: AbortSignal;
  /** Called with each content token during streaming (this is the final text). */
  onToken?: (token: string) => void;
  /** Called with each reasoning/thinking token during streaming (ephemeral, not in final result). */
  onReasoning?: (token: string) => void;
}

/**
 * Minimal HTTP client interface for dependency injection (testability).
 */
export interface HttpClient {
  request(
    url: URL,
    options: http.RequestOptions,
    callback: (res: http.IncomingMessage) => void
  ): http.ClientRequest;
}

const defaultHttpClient: HttpClient = {
  request(url: URL, options: http.RequestOptions, callback: (res: http.IncomingMessage) => void) {
    const mod = url.protocol === 'https:' ? https : http;
    return mod.request(url, options, callback);
  },
};

let httpClient: HttpClient = defaultHttpClient;

/**
 * Override the HTTP client used by processWithLLM / processWithLLMStream().
 * For testing only.
 */
export function _setHttpClient(client: HttpClient | null): void {
  httpClient = client ?? defaultHttpClient;
}

/**
 * Process transcribed text through an LLM for post-processing.
 * Ported from Swift LLMService.
 */
export async function processWithLLM(
  text: string,
  settings: LLMSettings,
  options: RequestOptions = {}
): Promise<string> {
  if (!settings.apiKey || settings.apiKey.trim() === '') {
    throw new LLMError(
      'API key is not configured. Open Settings to set it.',
      'NO_API_KEY'
    );
  }

  const baseURL = settings.apiBaseURL.replace(/\/+$/, '');
  let url: URL;
  try {
    url = new URL(`${baseURL}/v1/chat/completions`);
  } catch {
    throw new LLMError('Invalid API base URL. Check Settings.', 'INVALID_URL');
  }

  // Build messages array
  const messages: Array<{ role: string; content: string }> = [];
  const systemPrompt = settings.llmSystemPrompt.trim();
  if (systemPrompt) {
    messages.push({ role: 'system', content: systemPrompt });
  }
  // Wrap transcription in tags to clearly separate data from instructions
  messages.push({ role: 'user', content: `<transcription>${text}</transcription>` });

  const requestBody = JSON.stringify({
    model: settings.llmModelName,
    messages,
  });

  return retryWithBackoff(
    () => {
      if (options.signal?.aborted) {
        throw new LLMError('Request aborted', 'NETWORK_ERROR');
      }

      return doRequest(url, settings.apiKey, requestBody, options.signal);
    },
    {
      shouldRetry: (err) => {
        if (err instanceof LLMError) {
          // Non-retryable error codes
          if (
            err.code === 'NO_API_KEY' ||
            err.code === 'INVALID_URL' ||
            err.code === 'DECODING_ERROR'
          ) {
            return false;
          }
          // Non-retryable 4xx (except 408/429)
          if (
            err.code === 'API_ERROR' &&
            err.statusCode !== undefined &&
            err.statusCode >= 400 &&
            err.statusCode < 500 &&
            err.statusCode !== 408 &&
            err.statusCode !== 429
          ) {
            return false;
          }
        }
        return true;
      },
    }
  );
}

interface ChatCompletionResponse {
  choices: Array<{
    message: {
      content: string;
    };
  }>;
}

function doRequest(
  url: URL,
  apiKey: string,
  body: string,
  signal?: AbortSignal
): Promise<string> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) {
      reject(new LLMError('Request aborted', 'NETWORK_ERROR'));
      return;
    }

    const options: http.RequestOptions = {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(body),
      },
      timeout: LLM_TIMEOUT_MS,
    };

    const req = httpClient.request(url, options, (res) => {
      const chunks: Buffer[] = [];

      res.on('data', (chunk: Buffer) => {
        chunks.push(chunk);
      });

      res.on('end', () => {
        const data = Buffer.concat(chunks);
        const statusCode = res.statusCode ?? 0;

        if (statusCode === 200) {
          try {
            const parsed = JSON.parse(data.toString('utf-8')) as ChatCompletionResponse;
            const content = parsed.choices?.[0]?.message?.content;
            if (typeof content !== 'string') {
              reject(new LLMError('Failed to parse LLM response.', 'DECODING_ERROR'));
              return;
            }
            resolve(content.trim());
          } catch {
            reject(new LLMError('Failed to parse LLM response.', 'DECODING_ERROR'));
          }
          return;
        }

        const errorBody = data.toString('utf-8') || 'Unknown error';
        reject(
          new LLMError(
            `API error: HTTP ${statusCode}: ${errorBody}`,
            'API_ERROR',
            statusCode
          )
        );
      });

      res.on('error', (err) => {
        reject(new LLMError(`Network error: ${err.message}`, 'NETWORK_ERROR'));
      });
    });

    req.on('timeout', () => {
      req.destroy();
      reject(new LLMError('Request timed out', 'NETWORK_ERROR'));
    });

    req.on('error', (err) => {
      reject(new LLMError(`Network error: ${err.message}`, 'NETWORK_ERROR'));
    });

    const abortHandler = () => {
      req.destroy(new Error('Request aborted'));
      reject(new LLMError('Request aborted', 'NETWORK_ERROR'));
    };

    signal?.addEventListener('abort', abortHandler, { once: true });

    req.write(body);
    req.end();

    req.on('close', () => {
      signal?.removeEventListener('abort', abortHandler);
    });
  });
}

/* ------------------------------------------------------------------ */
/*  Streaming variant                                                  */
/* ------------------------------------------------------------------ */

interface SSEDelta {
  choices?: Array<{
    delta?: {
      content?: string;
      reasoning_content?: string;
      reasoning?: string;
    };
    finish_reason?: string | null;
  }>;
}

/**
 * Process transcribed text through an LLM with streaming (SSE).
 * Calls `options.onToken` for each token as it arrives.
 * Returns the fully-accumulated text when the stream ends.
 */
export async function processWithLLMStream(
  text: string,
  settings: LLMSettings,
  options: RequestOptions = {},
): Promise<string> {
  if (!settings.apiKey || settings.apiKey.trim() === '') {
    throw new LLMError(
      'API key is not configured. Open Settings to set it.',
      'NO_API_KEY',
    );
  }

  const baseURL = settings.apiBaseURL.replace(/\/+$/, '');
  let url: URL;
  try {
    url = new URL(`${baseURL}/v1/chat/completions`);
  } catch {
    throw new LLMError('Invalid API base URL. Check Settings.', 'INVALID_URL');
  }

  const messages: Array<{ role: string; content: string }> = [];
  const systemPrompt = settings.llmSystemPrompt.trim();
  if (systemPrompt) {
    messages.push({ role: 'system', content: systemPrompt });
  }
  messages.push({ role: 'user', content: `<transcription>${text}</transcription>` });

  const requestBody = JSON.stringify({
    model: settings.llmModelName,
    messages,
    stream: true,
  });

  return retryWithBackoff(
    () => {
      if (options.signal?.aborted) {
        throw new LLMError('Request aborted', 'NETWORK_ERROR');
      }
      return doStreamRequest(url, settings.apiKey, requestBody, options.signal, options.onToken, options.onReasoning);
    },
    {
      shouldRetry: (err) => {
        if (err instanceof LLMError) {
          if (
            err.code === 'NO_API_KEY' ||
            err.code === 'INVALID_URL' ||
            err.code === 'DECODING_ERROR'
          ) {
            return false;
          }
          if (
            err.code === 'API_ERROR' &&
            err.statusCode !== undefined &&
            err.statusCode >= 400 &&
            err.statusCode < 500 &&
            err.statusCode !== 408 &&
            err.statusCode !== 429
          ) {
            return false;
          }
        }
        return true;
      },
    },
  );
}

function doStreamRequest(
  url: URL,
  apiKey: string,
  body: string,
  signal?: AbortSignal,
  onToken?: (token: string) => void,
  onReasoning?: (token: string) => void,
): Promise<string> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) {
      reject(new LLMError('Request aborted', 'NETWORK_ERROR'));
      return;
    }

    const reqOptions: http.RequestOptions = {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(body),
      },
      timeout: LLM_TIMEOUT_MS,
    };

    const req = httpClient.request(url, reqOptions, (res) => {
      const statusCode = res.statusCode ?? 0;

      // Non-200: buffer error body and reject
      if (statusCode !== 200) {
        const errorChunks: Buffer[] = [];
        res.on('data', (chunk: Buffer) => errorChunks.push(chunk));
        res.on('end', () => {
          const errorBody = Buffer.concat(errorChunks).toString('utf-8') || 'Unknown error';
          reject(new LLMError(`API error: HTTP ${statusCode}: ${errorBody}`, 'API_ERROR', statusCode));
        });
        res.on('error', (err) => {
          reject(new LLMError(`Network error: ${err.message}`, 'NETWORK_ERROR'));
        });
        return;
      }

      // 200: parse SSE stream
      let accumulated = '';
      let sseBuffer = '';

      res.on('data', (chunk: Buffer) => {
        sseBuffer += chunk.toString('utf-8');

        // Process complete SSE lines
        const lines = sseBuffer.split('\n');
        // Keep the last (possibly incomplete) line in the buffer
        sseBuffer = lines.pop() ?? '';

        for (const line of lines) {
          const trimmed = line.trim();
          if (!trimmed || trimmed.startsWith(':')) continue; // comment or blank
          if (trimmed === 'data: [DONE]') continue;
          if (!trimmed.startsWith('data: ')) continue;

          const jsonStr = trimmed.slice(6); // strip 'data: '
          try {
            const parsed = JSON.parse(jsonStr) as SSEDelta;
            const delta = parsed.choices?.[0]?.delta;
            if (!delta) continue;

            // Reasoning/thinking tokens — ephemeral, shown in overlay but not in final result
            const reasoning = delta.reasoning_content ?? delta.reasoning ?? '';
            if (reasoning.length > 0) {
              onReasoning?.(reasoning);
            }

            // Content tokens — the actual output that gets pasted and saved
            const content = delta.content ?? '';
            if (content.length > 0) {
              accumulated += content;
              onToken?.(content);
            }
          } catch {
            // Ignore malformed SSE lines
          }
        }
      });

      res.on('end', () => {
        if (accumulated.length === 0) {
          reject(new LLMError('Empty streaming response from LLM.', 'DECODING_ERROR'));
          return;
        }
        resolve(accumulated.trim());
      });

      res.on('error', (err) => {
        reject(new LLMError(`Network error: ${err.message}`, 'NETWORK_ERROR'));
      });
    });

    req.on('timeout', () => {
      req.destroy();
      reject(new LLMError('Request timed out', 'NETWORK_ERROR'));
    });

    req.on('error', (err) => {
      reject(new LLMError(`Network error: ${err.message}`, 'NETWORK_ERROR'));
    });

    const abortHandler = () => {
      req.destroy(new Error('Request aborted'));
      reject(new LLMError('Request aborted', 'NETWORK_ERROR'));
    };

    signal?.addEventListener('abort', abortHandler, { once: true });

    req.write(body);
    req.end();

    req.on('close', () => {
      signal?.removeEventListener('abort', abortHandler);
    });
  });
}
